Test_Automation/library/danos_cli.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/python3
+# * Copyright (c) 2020-2021, Happiest Minds Technologies Limited Intellectual Property. All rights reserved.
+# *
+# * SPDX-License-Identifier: LGPL-2.1-only
+import time
+import vymgmt
+
+def config_set(ip, user, ps, cmd):
```

Comment:
I haven't used vymgmt before - but I can see a lot of duplication here.

Have you considered creating a connection object that creates the handles & login to the devices and cleans up afterward (logout) rather than doing it in each function in this file. This way you would only need to pass the handle to the device you wanted to configure. It would also speed up config as you wouldn't be creating a router object, logging in and out for each command you wanted to do.

Reply:
Hi, Trumpton,
Yes. I was expected this comment. Initially, tried constructing the python methods for different functions (like, connection, operational commands, and configuration commands, etc.) and used in the python scripts. But, the time taken for configuration with those methods was higher. During this time, found vymgmt option. Initially, during the development of scripts, had used a similar approach by creating handles as that one is the normal way one needs to be followed. But, for some reason, we could not re-use the handles from robot scripts and could not get a solution for that. I had raised issues also on that and they have not come back yet.
Hence, built python methods which are required for the different operations and used them. Also, taken care of exiting from the handle from each method.
On similar functions: The library is completely copied here. We have also automated a few other functionalities. Once those are also uploaded, those methods also will be used.
I would like to keep this as is until the solution is identified.

Reply:
Hi Dinesh,

On the reusing handles issue - this might be a possible solution...
https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#library-scope

If you create a danos_cli class in python and use the robot scoping,

`
class danos_cli():
    ROBOT_LIBRARY_SCOPE = 'GLOBAL'

   def __init__(self):
     blah
`

It will cause just a single object to be created for your test - thus the handles would be persistent. Without it I think you will find the handles are destroyed for each test.

Reply:
Thanks for this input Trumpton. I will explore this and if it is succeeded, will plan to implement it for the new scripts to start with.

Reply:
Hi Trumpton,
There are some issues with this method of ROBOT_LIBRARY_SCOPE as it is not behaving correctly when the sessions are used b/w test cases and suites. Tried with ROBOT_LIBRARY_SCOPE with TEST, SUITE, GLOBAL, and with the combination.
Finally, figured-out with robot keywords for generic functions and throughout the script, only one time the sessions to the router is opened and at the end, logout is used. Using 'alias' to switch b/w the sessions for a different router. I have used this method completely for the BGP script which is uploaded few minutes ago. Please have a look at the same and provide your feedback